solidity/contracts/liquidity-protection/LiquidityProtectionStore.sol

```diff
@@ -11,51 +11,73 @@ import "../utility/Utils.sol";
     using SafeMath for uint256;
 
-    struct ProtectedLiquidity {
+    uint256 private constant MAX_UINT128 = 2**128 - 1;
+    uint256 private constant MAX_UINT112 = 2**112 - 1;
+    uint256 private constant MAX_UINT32 = 2**32 - 1;
+
+    // the owner role is used to add values to the store, but it can't update them
+    bytes32 public constant ROLE_OWNER = keccak256("ROLE_OWNER");
+
+    // the seeder roles is used to seed the store with past values
+    bytes32 public constant ROLE_SEEDER = keccak256("ROLE_SEEDER");
+
+    struct Position {
         address provider; // liquidity provider
         uint256 index; // index in the provider liquidity ids array
         IDSToken poolToken; // pool token address
         IERC20Token reserveToken; // reserve token address
-        uint256 poolAmount; // pool token amount
-        uint256 reserveAmount; // reserve token amount
-        uint256 reserveRateN; // rate of 1 protected reserve token in units of the other reserve token (numerator)
-        uint256 reserveRateD; // rate of 1 protected reserve token in units of the other reserve token (denominator)
-        uint256 timestamp; // timestamp
+        uint128 poolAmount; // pool token amount
+        uint128 reserveAmount; // reserve token amount
+        uint256 reserveRateInfo; // reserve rate details:
+        // bits 0...111 represent the numerator of the rate between the protected reserve token and the other reserve token
+        // bits 111...223 represent the denominator of the rate between the protected reserve token and the other reserve token
+        // bits 224...255 represent the update-time of the rate between the protected reserve token and the other reserve token
+        // where `numerator / denominator` gives the worth of one protected reserve token in units of the other reserve token
     }
 
     struct LockedBalance {
         uint256 amount; // amount of network tokens
         uint256 expirationTime; // lock expiration time
     }
 
-    // protected liquidity by provider
-    uint256 private nextProtectedLiquidityId;
-    mapping(address => uint256[]) private protectedLiquidityIdsByProvider;
-    mapping(uint256 => ProtectedLiquidity) private protectedLiquidities;
+    // position by provider
+    uint256 private nextPositionId;
+    mapping(address => uint256[]) private positionIdsByProvider;
+    mapping(uint256 => Position) private positions;
 
     // user locked network token balances
     mapping(address => LockedBalance[]) private lockedBalances;
 
     // system balances
     mapping(IERC20Token => uint256) private systemBalances;
 
-    // total protected pool supplies / reserve amounts
-    mapping(IDSToken => uint256) private totalProtectedPoolAmounts;
-    mapping(IDSToken => mapping(IERC20Token => uint256)) private totalProtectedReserveAmounts;
+    // allows execution by the owner only
+    modifier ownerOnly {
```

Comment:
missing binary size optimizations that we use for modifiers + modifiers before ctor?

Reply:
We don't have a binary-size problem here.
This optimization is on the expanse of higher gas-cost during runtime, so we're probably better off without it.